mmv1/templates/terraform/constants/subnetwork.erb

```diff
@@ -17,3 +17,31 @@ func IsShrinkageIpCidr(_ context.Context, old, new, _ interface{}) bool {
 
 	return true
 }
+
+func sendSecondaryIpRangeIfEmptyDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
+	sendZero := diff.Get("send_secondary_ip_range_if_empty").(bool)
+	if !sendZero {
+		return nil
+	}
+
+	configValueIsEmpty := false
+	configSecondaryIpRange := diff.GetRawConfig().GetAttr("secondary_ip_range")
```

Comment:
Have you tested a case where the value of the field is unknown? Specifically, is the value considered `Null` in that case? One easy way to simulate unknowns is `local_file + data.local_file` I think.

Just a quick refresher, CustomizeDiff runs twice; once during plan, and once during apply immediately before calling [the internal SDK function] Apply on the resource. Different information can be present each time because interpolations are not resolved at plan time but are resolved at apply time. Terraform gets annoyed if known values change, typically returning an error, although to what degree in a messy case like this I'm not sure.

Because the field is O+C there are two ways it can be unknown:

- No entries are defined
- Entries are defined, but at least one of their values are unknown and we're in the plan phase

So it's possible we may need to report the value as _unknown_ during plan, and then set the known value of explicit zero during apply. I'm not entirely sure we can tell the two phases apart, either. There are GetRawPlan and NewValueKnown that may help provide more information.

Reply:
(we may also just decide to compromise and make it so that `send_secondary_ip_range_if_empty` overrides _all_ unknowns with a big 'ol warning)

Reply:
@SarahFrench icyi, we are pretty into the weeds on this stuff.

Reply:
I'm by no means an expert but there are methods for explicitly handling unknown values: [IsKnown](https://pkg.go.dev/github.com/hashicorp/go-cty@v1.4.1-0.20200414143053-d3edf31b6320/cty#Value.IsKnown), [IsWhollyKnown](https://pkg.go.dev/github.com/hashicorp/go-cty@v1.4.1-0.20200414143053-d3edf31b6320/cty#Value.IsWhollyKnown)

---

Sidebar that might be helpful: when I've looked into ways to allow unsetting of O+C fields I've looked at [this example](https://github.com/hashicorp/terraform-provider-aws/blob/b257779038f383ff3670f33a8637786ed1ce1673/internal/service/appstream/stack.go#L208-L240) from the AWS provider. This code detects if application_settings is in the state but isn't in the config, i.e. detects a scenario when the user has deleted the [application_settings block](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/appstream_stack#application_settings) completely from their config, and then sets a new 'zero' value for fields in that block. This code only works because application_settings is a top level field and is marked as Computed (requirements to use [schema#ResourceDiff.SetNew](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/schema#ResourceDiff.SetNew)). I couldn't find a solution for fields that don't meet those requirements. Because of this, I think your approach of adding a new field is correct for solving this type of problem specifically with the SDK.

Reply:
Thanks @SarahFrench, I'll take a look at that code - especially for handing unknown values

If I'm understanding it correctly, `application_settings` loses the ability to go from a Terraform configured value -> default to the API. It is assumed unsetting the field in an update means setting the empty value. The field can default to the API if initially created without being set in config.

@rileykarson what do you think of that model? Unsetting the field in an update is always assumed to be setting the empty value - which would eliminate the need for a virtual flag

Reply:
I think that'd be reasonable. Can the field still be set to empty on creation?

Reply:
If we ran `terraform apply` a second time after creation would we see a diff from the API default to empty?

Subnetwork is a multiactor case so it's less a priority but GKE cluster (at least, one of the fields) is a list of IPs allowed to access the cluster. For that case create-with-zero is a flow we pretty strongly want to support.

Reply:
I've confirmed that sending an empty `secondary_ip_range` list during create does not make the API actually return an empty list after creation. (gpaste/6269886252974080)

 Given this and the multiactor case, we should not need to send an empty `secondary_ip_range` after create.
 
 I've modified this customDiff to exit early if `IsKnown()` is false, similar to the previous example @SarahFrench linked.

Reply:
> I've confirmed that sending an empty secondary_ip_range list during create does not make the API actually return an empty list after creation.

I think I'm missing somethere here it returns null i.e. empty for lists in that paste.

> If we ran terraform apply a second time after creation would we see a diff from the API default to empty?

Sorry, this was me grounding too much on GKE. Not as relevant here as a new subnetwork should default to empty.

Reply:
Also flagging that the unknown nonzero value case has not been covered or I have not realised it has been covered anywhere yet! (apologies for two topics in one thread)

Reply:
> I think I'm missing somethere here it returns null i.e. empty for lists in that paste.

We send `"secondaryIpRanges": []` but don't get anything back during create. Not sending `secondaryIpRanges` will end up with the same effect on create. 

If something were to change `secondaryIpRanges` out of band and the user wanted to set it back to empty, the new flag will trigger an update on next apply.

Reply:
marking as resolved an continuing the 2nd topic in another thread